### ISSUE TITLE
SDK API call response log cannot be enabled as documented for .NET Core

### ISSUE THREAD

**Issue by fbouteruche** [aws/aws-sdk-net#1759]
## Description

As a .NET developer, I want to active SDK call response log to System.Diagnostics or to Console to log locally the responses to all SDK API calls when I'm debugging a .NET Core Web App.

I read carefully the documentation for .NET Core here https://docs.aws.amazon.com/sdk-for-net/latest/developer-guide/net-dg-config-netcore.html and added the following section to my appsettings.Development.json file :

```json  
"AWS": {
    "LogResponse": true,
    "LogMetrics": true,
    "DisableLogging":  false
  }
```
At this point, I except to see in the console or in the Debug Window of Visual Studio the logs of the responses to AWS API.

From my exploration of the SDK, I discovered that setting the following properties of the AWSConfigs object in addition to adding the appsettings file section activates the log:

```csharp
Amazon.AWSConfigs.LoggingConfig.LogResponses = Amazon.ResponseLoggingOption.Always;
Amazon.AWSConfigs.LoggingConfig.LogTo = Amazon.LoggingOptions.SystemDiagnostics;
Amazon.AWSConfigs.AddTraceListener("Amazon", new System.Diagnostics.ConsoleTraceListener());
```

Adding only 

```csharp
Amazon.AWSConfigs.AddTraceListener("Amazon", new System.Diagnostics.ConsoleTraceListener());
```

is not enough.

So is it a documentation issue or a code issue?

## Reproduction Steps

In the main branch of this repository https://github.com/fbouteruche/AWSSDKCallLogWebAppSample, you will find a sample that demonstrates that adding the above section will not output any logs in the console or in the Debug window.

In the branch https://github.com/fbouteruche/AWSSDKCallLogWebAppSample/tree/workaround-to-enable-local-log, you will find a sample of the proposed workaround.

### Logs

This issue is about log.

### Environment

* SDK Version: 
    * AWSSDK.Core 3.5.1.46
* Package Version: 
    * AWSSDK.Extensions.NETCore.Setup 3.3.101
    * AWSSDK.S3 3.5.6.2 

* OS Info: Windows 10
* Build Environment: Visual Studio 2019 16.8.2
* Targeted .NET Platform:  netcoreapp3.1

## Resolution
- [x] :wave: I can/would-like-to implement a fix for this problem myself

---

This is a :bug: bug-report

**Comment by ashishdhingra** [comment]
Hi @fbouteruche,

Thanks for posting the issue. 

As per API documentation for [Amazon.Runtime.ClientConfig](https://docs.aws.amazon.com/sdkfornet/v3/apidocs/items/Runtime/TClientConfig.html),  the settings `LogResponse` and `LogMetrics` are just the `boolean` flags. The configuration also needs other settings on where to log the output. Following references could be helpful:

- [Configuration Files Reference for AWS SDK for .NET](https://docs.aws.amazon.com/sdk-for-net/v3/developer-guide/net-dg-config-ref.html) 
- [Alternative Formatting for Metrics Data in .NET SDK Logs](https://aws.amazon.com/blogs/developer/alternative-formatting-for-metrics-data-in-net-sdk-logs/)
- [Retrieving Request Metrics from the AWS SDK for .NET](https://aws.amazon.com/blogs/developer/retrieving-request-metrics-from-the-aws-sdk-for-net/)

You are correctly setting the following properties of the AWSConfigs object to enable logging.

You may also refer [aws-logging-dotnet](https://github.com/aws/aws-logging-dotnet) which has plugins for popular .NET logging frameworks that integrate with Amazon Web Services. The plugins use the Amazon CloudWatch Logs service to write log data to a configured log group. [ASP.NET Core Logging](https://github.com/aws/aws-logging-dotnet#aspnet-core-logging) might be of interest to you.

Please review above information and let me know if this issue could be closed.

Thanks,
Ashish

**Comment by fbouteruche** [comment]
Hi @ashishdhingra 

Thanks for your quick answer and confirming that the additional steps are required.

However, I must point out that the link you provided are not relevant for .NET Core. A major change concerning configuration management has raised with the release of .NET Core as mentioned here https://aws.amazon.com/fr/blogs/developer/configuring-aws-sdk-with-net-core/.

This change led to the release of [AWSSDK.Extensions.NETCore.Setup](https://www.nuget.org/packages/AWSSDK.Extensions.NETCore.Setup).

Today, when you are developing a netcoreapp3.1web app, the following value are not read from the config file anymore: 
- Amazon.AWSConfigs.Logging
- Amazon.AWSConfigs.LogMetrics
- Amazon.AWSConfigs.ResponseLogging
- Amazon.AWSConfigs.LoggingConfig.LogMetrics
- Amazon.AWSConfigs.LoggingConfig.LogResponses
- Amazon.AWSConfigs.LoggingConfig.LogTo

This is mainly due to the fact that the GetConfig method from the AWSConfigs .netstandard implementation return null https://github.com/aws/aws-sdk-net/blob/e729b1958a37b62d433459ad41df202f544e09b6/sdk/src/Core/_netstandard/AWSConfigs.netstandard.cs#L38 .

So, while with .NET Framework, you can activate and configure log only through the configuration file, with .NET Core you are forced to mix both config file and code configuration which is a pity.

Concerning the use of aws-logging-dotnet, one doesn't necessarily want to send logs to CloudWatch, especially in early development stage when you are debugging on your own laptop. 

To sum up, I don't think you can close the issue. I think there is a documentation work needed here as well a fix for loading all the log configuration from the appsettings.json file in .NET Core.

**Comment by ashishdhingra** [comment]
Hi @fbouteruche,

Thanks for your inputs. I definitely see the need here to update the documentation. For reading from `aoosettings.json` file, I would need to discuss with developer on whether to treat this as a `bug` or `feature-request`.

Thanks,
Ashish

**Comment by dthomas-ss** [comment]
+1 for this this issue. We are experiencing the exact difficulties @fbouteruche

**Comment by shirlik** [comment]
+1 for this this issue. We are experiencing the exact difficulties @fbouteruche
when it is expected to be solved?

**Comment by techdragon** [comment]
Is there any indication this is on track to be fixed? The fact the "needs reproduction" label was only recently removed, doesn't fill me with confidence... And logging is an area where I personally have little tolerance for issues and friction.

**Comment by peterrsongg** [comment]
To enable logging in your .NET core application, you must inject the service client into .NET's dependency injection system. In NetCore the configuration is an instance level object which means we don't have a global way of accessing the configuration like we do in NetFramework, where the configuration is a global singleton. This is why you see that in `AWSConfigs.NetStandard` we just return null for `GetConfig`, because we expect the user to pass the `IConfiguration` to the SDK by using the NetCoreSetup package. To have the SDK read the appsettings.json file in NetCore, you have to add this line to your application
```
builder.Services.AddAWSService<IAmazonS3>();
```
Then, in your application you can then see that `Config.LogResponse` will reflect that which you set in `appsettings.Development.json`. Please let me know if that helps or you have any follow up questions
